File: ios/DJMediWallet/DJMediWallet/Core/WalletConfig.swift
//
//  WalletConfig.swift
//  DJMediWallet
//
//  Configuration for wallet initialization
//

import Foundation

/// Configuration options for wallet initialization
public struct WalletConfig {
    /// The service name for keychain storage (must not contain ":")
    public let serviceName: String
    
    /// The keychain access group for sharing credentials between apps (optional)
    public let accessGroup: String?
    
    /// Whether user authentication (biometric/passcode) is required for key operations
    public let userAuthenticationRequired: Bool
    
    /// Timeout in seconds after authentication before re-authentication is required (iOS only)
    public let authenticationTimeout: TimeInterval
    
    /// Whether to use Secure Enclave when available (iOS only)
    public let useSecureEnclaveWhenAvailable: Bool
    
    /// Trusted reader certificates for verification
    public let trustedReaderCertificates: [Data]?
    
    /// Default service name
    public static let defaultServiceName = "com.djmediwallet"
    
    /// Create wallet configuration
    /// - Parameters:
    ///   - serviceName: Service name for keychain (default: "com.djmediwallet")
    ///   - accessGroup: Keychain access group for app groups (optional)
    ///   - userAuthenticationRequired: Require biometric/passcode for operations (default: true)
    ///   - authenticationTimeout: Seconds before re-authentication required (default: 30)
    ///   - useSecureEnclaveWhenAvailable: Use Secure Enclave if available (default: true)
    ///   - trustedReaderCertificates: Trusted certificates for readers (optional)
    public init(
        serviceName: String = defaultServiceName,
        accessGroup: String? = nil,
        userAuthenticationRequired: Bool = true,
        authenticationTimeout: TimeInterval = 30,
        useSecureEnclaveWhenAvailable: Bool = true,
        trustedReaderCertificates: [Data]? = nil
    ) throws {
        // Validate service name doesn't contain ":"
        guard !serviceName.contains(":") else {
            throw WalletError.invalidConfiguration("Service name cannot contain ':' character")
        }
        
        self.serviceName = serviceName
        self.accessGroup = accessGroup
        self.userAuthenticationRequired = userAuthenticationRequired
        self.authenticationTimeout = authenticationTimeout
        self.useSecureEnclaveWhenAvailable = useSecureEnclaveWhenAvailable
        self.trustedReaderCertificates = trustedReaderCertificates
    }
    
    /// Default configuration
    public static var `default`: WalletConfig {
        return try! WalletConfig()
    }
}

// MARK: - Wallet Error Extension

extension WalletError {
    public static func invalidConfiguration(_ message: String) -> WalletError {
        return .unknown(NSError(domain: "WalletConfig", code: -1, userInfo: [NSLocalizedDescriptionKey: message]))
    }
}

File: android/DJMediWallet/app/src/main/java/com/djmediwallet/core/WalletConfig.kt
package com.djmediwallet.core

import java.util.*

/**
 * Configuration options for wallet initialization
 */
data class WalletConfig(
    /**
     * Service identifier for keystore (default: "DJMediWalletDeviceKey")
     */
    val serviceName: String = DEFAULT_SERVICE_NAME,
    
    /**
     * Whether user authentication (biometric/PIN) is required for key operations
     */
    val userAuthenticationRequired: Boolean = true,
    
    /**
     * Timeout in seconds after authentication before re-authentication is required
     * -1 means authentication is required for every use
     */
    val authenticationTimeoutSeconds: Int = 30,
    
    /**
     * Whether to use StrongBox when available (Android 9+)
     */
    val useStrongBoxWhenAvailable: Boolean = true,
    
    /**
     * Trusted reader certificates for verification
     */
    val trustedReaderCertificates: List<ByteArray>? = null
) {
    
    init {
        // Validate service name doesn't contain ":"
        require(!serviceName.contains(":")) {
            "Service name cannot contain ':' character"
        }
        
        require(authenticationTimeoutSeconds == -1 || authenticationTimeoutSeconds > 0) {
            "Authentication timeout must be -1 (always) or positive"
        }
    }
    
    companion object {
        const val DEFAULT_SERVICE_NAME = "DJMediWalletDeviceKey"
        
        /**
         * Default configuration with standard security settings
         */
        val DEFAULT = WalletConfig()
        
        /**
         * Debug configuration with authentication disabled
         */
        val DEBUG = WalletConfig(
            userAuthenticationRequired = false,
            authenticationTimeoutSeconds = -1
        )
    }
    
    /**
     * Builder for creating WalletConfig with fluent API
     */
    class Builder {
        private var serviceName: String = DEFAULT_SERVICE_NAME
        private var userAuthenticationRequired: Boolean = true
        private var authenticationTimeoutSeconds: Int = 30
        private var useStrongBoxWhenAvailable: Boolean = true
        private var trustedReaderCertificates: List<ByteArray>? = null
        
        fun serviceName(name: String) = apply { this.serviceName = name }
        
        fun userAuthenticationRequired(required: Boolean) = apply { 
            this.userAuthenticationRequired = required 
        }
        
        fun authenticationTimeoutSeconds(seconds: Int) = apply { 
            this.authenticationTimeoutSeconds = seconds 
        }
        
        fun useStrongBox(use: Boolean) = apply { 
            this.useStrongBoxWhenAvailable = use 
        }
        
        fun trustedReaderCertificates(certificates: List<ByteArray>?) = apply { 
            this.trustedReaderCertificates = certificates 
        }
        
        fun build(): WalletConfig {
            return WalletConfig(
                serviceName = serviceName,
                userAuthenticationRequired = userAuthenticationRequired,
                authenticationTimeoutSeconds = authenticationTimeoutSeconds,
                useStrongBoxWhenAvailable = useStrongBoxWhenAvailable,
                trustedReaderCertificates = trustedReaderCertificates
            )
        }
    }
}
